### ISSUE TITLE
How to handle hard links?

### ISSUE THREAD

**Issue by zevv** [zevv/duc#6]
What would be the proper way to handle counting size with hard links? 

Double counting hard links:

- pro: simple implementation
- pro: shows where the data is
- con: totals do not add up

Single counting of hard links:

- pro: totals add up
- con: big files show up in one place, but not in the other.
- con: slower indexing and more memory usage because a list of visited inodes must be kept.

**Comment by zevv** [comment]
As of version 1c898c7 duc has support for counting hard links only once. Use the -H/--check-hard-links option on `duc index`.

The current implementation (based on the tokyo cabinet in-memory hash map) is not very memory efficient, it takes approx 64 bytes for each scanned inode in the hash table. Scanning 15 million files will take up a gigabyte of memory, which I think is not really acceptable.

**Comment by asmartin** [comment]
Excellent, I'll give this a try and report back my results!

**Comment by l8gravely** [comment]
>>>>> "zevv" == zevv  <notifications@github.com> writes:

zevv> As of version 1c898c7 duc has support for counting hard links
zevv> only once. Use the -H/--check-hard-links option on duc index.
zevv> The current implementation (based on the tokyo cabinet in-memory
zevv> hash map) is not very memory efficient, it takes approx 64 bytes
zevv> for each scanned inode in the hash table. Scanning 15 million
zevv> files will take up a gigabyte of memory, which I think is not
zevv> really acceptable.

One idea might be to just ignore hardlinks below a certain size, since
I figure they aren't really used all that much.  Also, a simple B-tree
might be a smarter model since you won't find that many and they will
all have the same Inode number, so just keeping track of the found
inodes will be enough.

And I'd then just assign the cost to the first directory it's found
in, which is no worse than any other scheme I think.  

Now, if the file that's hard linked is over say 1mb in size (or some
percentage of total file system space) or the usage of all the
hardlinks (which is counted in each inode) sums up to a large
percentage, then maybe something special gets done, or a warning is
printed.

But even if someone deletes a hardlink, the space won't come back
until they're all deleted, so they're not *that* useful to target
IMHO.  Am I making sense here?

John

**Comment by zevv** [comment]
* On 2015-04-22 17:25:01 +0200, John wrote:
 
> >>>>> "zevv" == zevv  <notifications@github.com> writes:
> 
> zevv> As of version 1c898c7 duc has support for counting hard links
> zevv> only once. Use the -H/--check-hard-links option on duc index.
> zevv> The current implementation (based on the tokyo cabinet in-memory
> zevv> hash map) is not very memory efficient, it takes approx 64 bytes
> zevv> for each scanned inode in the hash table. Scanning 15 million
> zevv> files will take up a gigabyte of memory, which I think is not
> zevv> really acceptable.
> 
> One idea might be to just ignore hardlinks below a certain size, since
> I figure they aren't really used all that much.

I don't think we should do that. We should either count hard links once, or
count the total, but now show fuzzy behaviour.

> Also, a simple B-tree might be a smarter model since you won't find
> that many and they will all have the same Inode number, so just
> keeping track of the found inodes will be enough.

But keeping track of the found inodes is exactly where the problem lies.
Consider my test system: I have a incremental backup which depends on
hard links, and has currently 250 million files on it. ino_t is 64 bits,
which is 8 bytes. Assuming worst case that each file has exactly 1 hard
link (so 2 duplicates of each file) will result in 125 million unique
inodes which need to be kept track of. 125e6 * 8 = 1 GB of RAM. That's
for the inodes only, without any overhead of the data structure used to
keep track of them.

My goal is to find a reasonable data structure to hold this data which
does not cause to much memory bloat. For now the tokyocabinet in-memory
hash will do, but since Duc has no other bottlenecks in scaling at this
time, it would be a pity if keeping track of hard links would be the
feature to cause problems.

As a fallback mechanism we could add a secondary store for keeping
track of inodes on disk, using a normal tokyocabinet db. We allow the
in-memory database to take up to a certain amount of system memory,
and as soon as this is exceeded the bookkeeping goes to an on-disk
database instead. Not hard to implement.

-- 
:wq
^X^Cy^K^X^C^C^C^C

**Comment by zevv** [comment]
The current implementation is not as bad at it seemed, it performs well enough on the tests I did so far. I'm closing this ticket to clean up the list, if anyone feels the current hardlink handling is not good enough feel free to reopen.
